File: interactive.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

# define STB_IMAGE_IMPLEMENTATION
#include "stb_image/stb_image.h"
#define STB_IMAGE_WRITE_IMPLEMENTATION
#include "stb_image/stb_image_write.h"

#include "filters.c"

int main() {
    char filename[1024] = "";
    printf("Name of image you would like to open: ");
    fgets(filename, 1024, stdin);
    filename[strlen(filename)-1] = 0;

    int width, height, channels;
    unsigned char *image = stbi_load(filename, &width, &height, &channels, 0);
    size_t size_of_image = width * height * channels;
    
    if (!image) {
        printf("Invalid image name or unable to load image.\n");
        return 1;
    }

    int choice;
    printf("Filter options: \n" \
           "1. Dark lines filter\n" \
           "2. Red filter\n" \
           "3. Empty filter\n" \
           "4. Green filter\n" \
           "5. Blue turquise filter\n" \
           "6. Broken screen filter\n" \
           "7. Reduced opacity filter\n" \
           "8. Dark lines filter\n" \
           "9. Exit program\n"
           "10. Grey Filter\n"
           "Select a number: ");
    scanf("%d", &choice);
    getchar();

    printf("selected: %d\n", choice);

    int opacity_level;
    bool repeat = true;
    bool exit = false;
    while (repeat) {
        switch (choice) {
            case 1:
                increase_shade_filter(size_of_image, image);
                repeat = false;
                break;
            case 2:
                red_filter(size_of_image, image);
                repeat = false;
                break;
            case 3:
                empty_filter(size_of_image, image);
                repeat = false;
                break;
            case 4:
                green_filter(size_of_image, image);
                repeat = false;
                break;
            case 5:
                blue_turquise_filter(size_of_image, image);
                repeat = false;
                break;
            case 6:
                broken_screen_filter(size_of_image, image);
                repeat = false;
                break;
            case 7:
                printf("Enter opacity level: ");
                scanf("%d", &opacity_level);
                getchar();
                reduced_opacity_filter(size_of_image, image, opacity_level);
                repeat = false;
                break;
            case 8:
                brighter_filter(size_of_image, image);
                repeat = false;
                break;
            case 9:
                exit = true;
                repeat = false;
                break;
            case 10:
                repeat = false;
                exit = true;
                greyscale_filter(size_of_image, width, height, channels, image);
            default:
                printf("Incorrect input.\n");
                repeat = true;
                break;
        }
    }

    if (exit) return 1;
    stbi_write_png("output_file.png", width, height, channels, image, width * channels);
    stbi_image_free(image);
    return 0;
}

File: interactive.py
from PIL import Image, ImageFilter
from utilities import *
from filters import *


def print_options():
    print("1. Greyscale filter")
    print("2. Green filter")
    print("3. Alternate greyscale filter")
    print("4. Blue filter")
    print("5. Dark blue filter")
    print("6. Yellow filter")
    print("7. Reduce opacity filter")
    print("8. Slight blur 3x3 filter")
    print("9. Slight blur 4x3 filter")
    print("10. Blur filter")
    print("11. Luminosity filter")
    print("12. Contrast filter")
    print("13. Crop image filter")
    print("14. Slight blur filter")
    print("15. Sharpen filter")
    print("16. Unsharpen mask filter")
    print("17. Edge detection filter")
    print("18. Exit program")


repeat = True
while (repeat):
    repeat_opening_image = True
    while(repeat_opening_image):
        try:
            image_name = input("Enter the image file name with the extension: ")
            image = Image.open(image_name)
            image = image.convert('RGB')
            repeat_opening_image = False
        except:
            print("File unable to open.")
            repeat_opening_image = True

    print_options()
    selection = int(input("Enter a number to select an option: "))
    
    if (selection == 1):
        greyscale_filter(image)
        repeat = False
    elif (selection == 2):
        green_filter(image)
        repeat = False
    elif (selection == 3):
        greyscale_alternative_filter(image)
        repeat = False
    elif (selection == 4):
        blue_filter(image)
        repeat = False
    elif (selection == 5):
        dark_blue_filter(image)
        repeat = False
    elif (selection == 6):
        yellow_filter(image)
        repeat = False
    elif (selection == 7):
        level_of_opacity = input("Level of opacity: ")
        reduce_opacity_filter(image, level_of_opacity)
        repeat = False
    elif (selection == 8):
        blur_filter_3x3(image)
        repeat = False
    elif (selection == 9):
        blur_filter_4x3(image)
        repeat = False
    elif (selection == 10):
        level_of_blur = input("Enter level of blur from 1-100: ")
        blur_lib_filter(image, level_of_blur)
        repeat = False
    elif (selection == 11):
        level_of_luminosity = input("Level of luminosity: ")
        luminosity_filter(image, level_of_luminosity)
        repeat = False
    elif (selection == 12):
        contrast_filter(image)
        repeat = False
    elif (selection == 13):
        x = input("Enter upper x-coordinate: ")
        y = input("Enter upper y-coordinate: ")
        upper_left_coordinates = (x, y)
        x = input("Enter lower x-coordinate: ")
        y = input("Enter lower y-coordinate: ")
        bottom_right_coordinates = (x, y)
        crop_filter(image, upper_left_coordinates, bottom_right_coordinates)
        repeat = False
    elif (selection == 14):
        blur_filter(image)
        repeat = False
    elif (selection == 15):
        sharpen_filter(image)
        repeat = False
    elif (selection == 16):
        unsharpen_mask_filter(image)
        repeat = False
    elif (selection == 17):
        edge_detection_filter(image)
        repeat = False
    elif (selection == 18):
        repeat = False
        exit()
    else:
        print("Invalid option.")
        repeat = True

